### DISCUSSION TITLE
Parsing callback SAML data

### DISCUSSION THREAD

**Discussion by joelwass** [bergie/passport-saml#644] [Q&A]
So i'm a little confused on how the callback function should look. Should the call back function be using the passport saml middleware? As pictured below and in your ReadMe? 
![screen shot 2016-04-27 at 5 16 59 pm](https://cloud.githubusercontent.com/assets/7344422/14868529/f4c975f0-0c9b-11e6-8c50-d776f60487d5.png)

or should the code be structured more like this?
![screen shot 2016-04-27 at 5 18 14 pm](https://cloud.githubusercontent.com/assets/7344422/14868540/0ec500fa-0c9c-11e6-8dc8-6a5e2d63e3b2.png)

The issue is -- i'm getting an error with the code in the first image (your provided example code) because the post request is attempting to send another authenticate saml request as I can see in my SAML browser plugin (with an invalid NameId of course), where as if i use the code in the second image, i can get a response back but the request isn't being parsed for me.. and I'm having to parse it myself in a very ugly manner.

Thanks in advance

**Comment by artis3n** [comment]
I assume the error you see is the same one I see from the callback route as described in the first picture. The error is `TypeError: Converting circular structure to JSON`. Seems to be at `at validateCallback (<project>/node_modules/passport-saml/lib/passport-saml/strategy.js:59:14)`.

At line 59 in `strategy.js`, we see the call `self._verify(req, profile, verified);`. There I lose the path of the logic.

Can I ask how you are parsing the request yourself? The project doesn't appear to be being maintained, yet we're stuck with it.

**Comment by EdouardPauw** [comment]
Hello,

I have kind of the same issue. After being redirected to my POST callback uri, this error is triggered : 
`"Converting circular structure to JSON\n    --> starting at object with constructor 'Socket'\n    |     property 'parser' -> object with constructor 'HTTPParser'\n    --- property 'socket' closes the circle"`

I tried to add some logs in the `verify`, `serializeUser` and `deserializeUser` functions but none are triggered.

Does any one have a solution ? I can't figure out what I'm doing wrong .. :'(

**Reply by srd90** [reply]
@EdouardPauw You could have provided some sample code of your application and/or full stack trace.

Are you trying to `stringify` directly or indirectly `req` object. Maybe you have some `log(JSON.stringify(req))` or something like that somewhere.

Have you tried some search engine?
https://www.google.com/search?q=express+Converting+circular+structure+to+JSON+socket+httpparser
Some highlights from the results:
1. https://github.com/expressjs/express/issues/4311#issuecomment-641983696
2. https://github.com/expressjs/express/issues/4311#issuecomment-950149333
3. https://stackoverflow.com/questions/66450846/nodejs-express-typeerror-converting-circular-structure-to-json

**Reply by EdouardPauw** [reply]
Thx a lot !
I succed to find the stack trace and indeed the issue was raised when we catched the error and tried to log the error and the request.
Now I cas see the real error : `Invalid signature` :)
